### PR TITLE
Support Globus SDK 3.x and other Globus improvements

### PR DESCRIPTION
**Pull request by Cadair**
Direct link to the changed doc page: https://dkistdc-public-documentation--136.com.readthedocs.build/projects/python-tools/en/136/guide/net.html

### CODE REVIEW

File: dkist/net/globus/transfer.py [hunk 1]
```diff
@@ -186,37 +201,124 @@ def watch_transfer_progress(task_id, tfr_client, poll_interval=5,
...
+    Transfer the files given in file_list to the path on ``destination_endpoint``.
+
+    Parameters
+    ----------
+    file_list
+
+    recursive
+
+    destination_path
+
```
Comment (eigenbrot) [inline dkist/net/globus/transfer.py:280]: Descriptions, please!

File: dkist/net/globus/transfer.py [hunk 2]
```diff
@@ -186,37 +201,124 @@ def watch_transfer_progress(task_id, tfr_client, poll_interval=5,
...
+        A unique specifier for a Globus endpoint. If `None` a local
+        endpoint will be used if it can be found, otherwise an error will
+        be raised. See `~dkist.net.globus.get_endpoint_id` for valid
+        endpoint specifiers.
+
+    progress
+        If `True` status information and a progress bar will be displayed
+        while waiting for the transfer to complete.
+        If ``progress="verbose"`` then all globus events generated during the
+        transfer will be shown (by default only error messages are shown.)
```
Comment (eigenbrot) [inline dkist/net/globus/transfer.py:291]: I think a sentence at the end saying something like "This keyword only matters if ``wait=True`` will help ensure people understand this behavior.

File: dkist/net/helpers.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    dataset_id = dataset["Dataset ID"]
+    proposal_id = dataset["Primary Proposal ID"]
+    bucket = dataset["Storage Bucket"]
+
+    destination_path = Path(path) / proposal_id
+
+    file_list = [Path(conf.dataset_path.format(
+        datasetId=dataset_id,
+        primaryProposalId=proposal_id,
+        bucket=bucket
```
Comment (eigenbrot) [inline dkist/net/helpers.py:85]: This usage constrains the pattern in `conf.dataset_path` and so if that path ever changes you'll need to update these lines as well. Is there a way to keep all of that logic in a single place?

File: dkist/net/helpers.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
+"""
```
Comment (eigenbrot) [inline dkist/net/helpers.py:1]: Not to be That Guy, but "helpers.py" is a guaranteed terrible name for a module.

File: docs/guide/net.rst [hunk 5]
```diff
@@ -3,19 +3,20 @@
 Searching and Downloading Level One Data
 ========================================
 
-The DKIST Data Center provides a search interface for searching all level 1 datasets and makes ASDF files available to download over HTTP, while FITS files are transferred via Globus.
-The DKIST Python tools makes use of sunpy's pluggable search and download API, which is described in :ref:`sunpy:fido_guide`.
+The DKIST Data Center provides a search interface for searching for level one datasets.
+This means you can search for whole collections of FITS files, each with associated an associated ASDF metadata file and other ancillary files.
```
Comment (eigenbrot) [inline docs/guide/net.rst:7]: Too many "associated"s

Comment (eigenbrot) [inline docs/guide/net.rst:7]: What would be an example of an "ancillary" file? Movies? Quality reports?

File: docs/guide/net.rst [hunk 6]
```diff
@@ -70,19 +71,115 @@ This will return a list-like object which contains one or more filepaths to the
...
+
+The `dkist` package provides two different interfaces for starting a Globus transfer.
+One is `~dkist.net.transfer_whole_dataset` which will stage the whole dataset directory for download.
+The other is the `~dkist.io.FileManager.download` method of ``Dataset.files``, this method allows you to only download some FITS files based on the slicing of a dataset object.
+
+
+Downloading Files from a ``Dataset``
+####################################
+
+The most flexible way to download FITS files is first having loaded the dataset from an ASDF file.
```
Comment (eigenbrot) [inline docs/guide/net.rst:86]: "...first having loaded the dataset..." is awkward; I'm not even sure what it's trying to say. Would "...to first load the dataset from an ASDF file." correct?

File: docs/guide/net.rst [hunk 7]
```diff
@@ -70,19 +71,115 @@ This will return a list-like object which contains one or more filepaths to the
...
+This workflow enables you to choose what FITS files you want to transfer based on the metadata of the dataset, for example, the wavelength or stokes profile.
```
Comment (eigenbrot) [inline docs/guide/net.rst:87]: Replace with "...dataset (for example, ...)"

File: docs/guide/net.rst [hunk 8]
```diff
@@ -70,19 +71,115 @@ This will return a list-like object which contains one or more filepaths to the
...
+
+The first step in this workflow is to construct a `dkist.Dataset` object from an already downloaded ASDF.
+
+How to do this is detailed in the next section :ref:`loadinglevel1data`, but a quick example, following on from the ``Fido.fetch`` call above:
```
Comment (eigenbrot) [inline docs/guide/net.rst:91]: need comma after "section"

File: docs/guide/net.rst [hunk 9]
```diff
@@ -70,19 +71,115 @@ This will return a list-like object which contains one or more filepaths to the
...
+For example:
+
+.. code-block:: python
+
+   ds.files.basepath = "~/data/dkist/BCDEF"
+
+.. note::
+
+   To transfer files to your local computer you will need a running instance of the `Globus Connect Personal <https://www.globus.org/globus-connect-personal>`__ software.
+   All the following documentation assumes you have this running and wish to transfer files using GCP to the machine where your Python session is running.
```
Comment (eigenbrot) [inline docs/guide/net.rst:115]: Either need to define GCP in the line above or just spell it out here.

File: docs/guide/net.rst [hunk 10]
```diff
@@ -70,19 +71,115 @@ This will return a list-like object which contains one or more filepaths to the
...
 
 .. code-block:: python
 
    ds.files.download()
 
-If this is the first time you have run this method a login page should open in your webbrowser to authenticate with the Globus system.
-Without overriding any of the default arguments calling this method **expects you to already have a running Globus personal endpoint** local to the Python session.
-If you wish to transfer the dataset to a non-local endpoint, you can do so by setting the ``destination_endpoint=`` keyword argument.
+If this is the first time you have run this method, or your authentication has expired, a login page should open in your webbrowser to authenticate with the Globus system.
+By default this call will download all the FITS files to the current value of ``ds.files.basepath``, i.e. next to the loaded ASDF file.
```
Comment (eigenbrot) [inline docs/guide/net.rst:125]: "i.e., next to the loaded ASDF file.": Based on lines 105 above my read is that ``ds.files.basepath`` isn't _always_ next to the ASDF file. Is that right? If so maybe text like "...basepath`` (defualts to the same directory as the loaded ASDF file)." would be more correct.

File: docs/guide/net.rst [hunk 11]
```diff
@@ -70,19 +71,115 @@ This will return a list-like object which contains one or more filepaths to the
...
+   results
+
+If we wanted to transfer all of the datasets returned by this search we could pass the results object to `~.transfer_whole_dataset`:
+
+.. code-block:: python
+
+    transfer_whole_dataset(results["dkist"])
+
+Note how we have to extract the ``"dkist"`` table from the `~sunpy.net.fido_factory.UnifiedResponse` object, as `.transfer_whole_dataset` only operates on DKIST datasets.
+This will iterate over each dataset in the results and transfer them one-by-one showing a progress bar for each one.
```
Comment (eigenbrot) [inline docs/guide/net.rst:171]: I'm confused by the pluralization here. The section title and function name of `transfer_whole_dataset` implies that only a single dataset will be downloaded, but here you say that it actually iterates over multiple datasets. What gives?

**Review by eigenbrot — CHANGES_REQUESTED** [review]
In addition to the in-line comments I'll also add that type-hints in the function definitions would be very useful and appreciated.

File: dkist/io/file_manager.py [hunk 12]
```diff
@@ -245,29 +266,28 @@ def download(self, path="/~/", destination_endpoint=None, progress=True):
...
-        destination_path = Path(path) / inv['primaryProposalId'] / inv['datasetId']
+        base_path = Path(net_conf.dataset_path.format(**inv))
+        destination_path = path or self.basepath or "/~/"
+        destination_path = Path(destination_path).as_posix()
+        destination_path = Path(destination_path.format(**inv))
 
+        # TODO: If we are transferring the whole dataset then we should use the
+        # directory not the list of all the files in it.
         file_list = [base_path / fn for fn in self.filenames]
         file_list.append(Path("/") / inv['bucket'] / inv['asdfObjectKey'])
```
Comment (Hyzerputt) [inline dkist/io/file_manager.py:277]: There is a browse movie and quality report in the same place as the asdf.  I concur with the directory transfer TODO, but in the mean time you might want to consider appending those to the file list like you do the asdf.

File: docs/guide/net.rst (hunk 11 above)
Comment (Hyzerputt) [reply docs/guide/net.rst:171]: Looks like the support for multiple datasets as an argument type to `transfer_whole_dataset` came later.  I think the outcome is that this block of text is fine but the name of `transfer_whole_dataset` and it's doc string need updating.

File: dkist/dataset/dataset.py [hunk 13]
```diff
@@ -190,7 +190,9 @@ def from_asdf(cls, filepath):
             with resources.path("dkist.io", "level_1_dataset_schema.yaml") as schema_path:
                 with asdf.open(filepath, custom_schema=schema_path.as_posix(),
                                lazy_load=False, copy_arrays=True) as ff:
-                    return ff.tree['dataset']
+                    ds = ff.tree['dataset']
+                    ds.files.basepath = base_path
```
[the remark this answers is not in the document]
Comment (Cadair) [reply dkist/dataset/dataset.py:194]: This is fixing a bug that when loading an asdf file it didn't assume the FITS were in the same directory.

File: dkist/io/file_manager.py [hunk 14]
```diff
@@ -245,29 +266,28 @@ def download(self, path="/~/", destination_endpoint=None, progress=True):
...
+        _orchestrate_transfer_task(file_list,
+                                   recursive=False,
+                                   destination_path=destination_path,
+                                   destination_endpoint=destination_endpoint,
+                                   progress=progress,
+                                   wait=wait)
 
         if is_local:
             local_destination = destination_path.relative_to("/").expanduser()
+            if local_destination.root == "":
```
[the remark this answers is not in the document]
Comment (Cadair) [reply dkist/io/file_manager.py:291]: This is to work around this: `Path("/data").relative_to("/").root`

File: dkist/net/helpers.py (hunk 3 above)
Comment (Cadair) [reply dkist/net/helpers.py:85]: Yeah, I know, but I can't think of a way out of it shy of a large refactoring of the Fido client.

**Comment by Cadair** [comment]
> type-hints in the function definitions would be very useful and appreciated

I am adding them as I write new functions but am not adding them back to old things at the moment. Frankly until I get them rendering nicer than [this horror show](https://dkistdc-public-documentation--136.com.readthedocs.build/projects/python-tools/en/136/api/dkist.net.transfer_complete_datasets.html#dkist.net.transfer_complete_datasets) I am unconvinced they are a net-positive for most users.

File: docs/guide/net.rst (hunk 11 above)
Comment (Cadair) [reply docs/guide/net.rst:171]: I have renamed the function and fixed up the docstring.

**Pull request merged**